File: iOS/AnalyticsManager/AnalyticsManager.swift
//
//  AnalyticsManager.swift
//  SwiftiDate
//
//  Created by 游哲維 on 2025/2/28.
//

import Foundation
// 如果要整合 Firebase Analytics:
import FirebaseAnalytics
import Mixpanel

class AnalyticsManager: AnalyticsManagerProtocol {
    static var shared = AnalyticsManager()

    private init() {
        // 若需要在這裡做額外初始化可加，如 FirebaseApp.configure() 等
        
        // 初始化 Mixpanel
        Mixpanel.initialize(token: "c33a19b0b1c17db46731337f2bc233da", trackAutomaticEvents: true)
    }
    
    /// 追蹤事件的統一方法
    @objc dynamic func trackEvent(_ eventName: String, parameters: [String: Any]? = nil) {
        // Firebase
        Analytics.logEvent(eventName, parameters: parameters)
        
        // Mixpanel
        let mixpanelProps = convertToMixpanelProperties(parameters)
        Mixpanel.mainInstance().track(event: eventName, properties: mixpanelProps)
        
        print("Tracking Event: \(eventName), params: \(parameters ?? [:])")
    }
    
    func convertToMixpanelProperties(_ dictionary: [String: Any]?) -> Properties {
        guard let dictionary = dictionary else { return [:] }

        var result: Properties = [:]
        for (key, value) in dictionary {
            // 檢查 value 是否符合 MixpanelType (String, Int, Double, Bool, Date, etc.)
            if let v = value as? MixpanelType {
                result[key] = v
            } else {
                // 如果是其他型別，視需求決定要 skip 還是做某種轉換
                // 這裡示範直接略過
                print("Warning: Skip \(key) because \(value) is not MixpanelType")
            }
        }
        return result
    }
}

File: flutter/swiftidate/lib/analytics/analytics_manager.dart
import 'package:firebase_analytics/firebase_analytics.dart';
import 'package:mixpanel_flutter/mixpanel_flutter.dart';
import 'analytics_manager_protocol.dart';

class AnalyticsManager implements AnalyticsManagerProtocol {
  // 單例
  static final AnalyticsManager shared = AnalyticsManager._internal();
  AnalyticsManager._internal();

  // FirebaseAnalytics 實例
  final FirebaseAnalytics _firebaseAnalytics = FirebaseAnalytics.instance;

  // Mixpanel 實例，需在初始化時設定 Token
  Mixpanel? _mixpanel;

  // 初始化 Mixpanel，需要在應用程式啟動時呼叫一次
  Future<void> initializeMixpanel(String token) async {
    _mixpanel = await Mixpanel.init(token, optOutTrackingDefault: false);
  }

  @override
  Future<void> trackEvent(String eventName, {Map<String, dynamic>? parameters}) async {
    // Firebase Analytics
    // 注意：Firebase Analytics 只接受 Map<String, Object?>
    final firebaseParams = <String, Object?>{};
    if (parameters != null) {
      for (final entry in parameters.entries) {
        // 只取能轉成 Firebase Analytics 支援的類型
        if (entry.value is String ||
            entry.value is num ||
            entry.value is bool) {
          firebaseParams[entry.key] = entry.value as Object;
        } else {
          // 其他型別可根據需求做轉換或直接略過
          print("Warning: Skip key=${entry.key} because value is not supported in Firebase Analytics");
        }
      }
    }
    await _firebaseAnalytics.logEvent(name: eventName, parameters: firebaseParams);

    // Mixpanel
    if (_mixpanel != null) {
      // Mixpanel 的 track 也需要 Map<String, dynamic>
      final mixpanelProps = <String, dynamic>{};
      if (parameters != null) {
        for (final entry in parameters.entries) {
          // Mixpanel 支援 string, num, bool, DateTime 等
          if (entry.value is String ||
              entry.value is num ||
              entry.value is bool ||
              entry.value is DateTime) {
            mixpanelProps[entry.key] = entry.value;
          } else {
            print("Warning: Skip key=${entry.key} in Mixpanel because value is not supported");
          }
        }
      }
      _mixpanel!.track(eventName, properties: mixpanelProps);
    } else {
      print("Warning: Mixpanel not initialized. Please call initializeMixpanel() first.");
    }

    // 在 Console 輸出
    print("Tracking Event: $eventName, params: $parameters");
  }
}
